### PR TITLE
Add jdbc upsert writer

### PR DESCRIPTION
**Pull request by wangqinghuan**
Hi, Spark dataframe does only support append or overwrite data. I implement a  simple JdbcUpsertWriter to upsert data in relation database. It will check whether the  dataframe row exists in database. If the row exists,generates an UPDATE statement and if row does not exists, creates an INSERT statement.

### CODE REVIEW

File: src/main/scala/com/yotpo/metorikku/output/writers/jdbc/JDBCUpsertWriter.scala [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+  }
+
+  def doUpdate(conn: Connection, updateStmt: PreparedStatement, updateFields: Array[String], row: Row): Unit = {
+    for (i <- 1 to updateFields.size) {
+      addValueToStatement(row.getAs(updateFields(i - 1)), updateStmt, i)
+    }
+    updateStmt.addBatch();
+  }
+
+  def addValueToStatement(v: Any, stmt: PreparedStatement, i: Int): Unit = {
```
Comment (RonBarabash) [inline src/main/scala/com/yotpo/metorikku/output/writers/jdbc/JDBCUpsertWriter.scala]: isnt it the same code from jdbc query? why the duplication?

File: src/main/scala/com/yotpo/metorikku/output/writers/jdbc/JDBCUpsertWriter.scala [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+      case v: BinaryType => stmt.setBytes(i, v.asInstanceOf[Array[Byte]])
+      case v: ArrayType => stmt.setString(i, v.asInstanceOf[ArrayType].json)
+      case v: MapType => stmt.setString(i, v.asInstanceOf[MapType].json)
+      case v: StructType => stmt.setString(i, v.asInstanceOf[StructType].json)
+      // NULLs
+      case _ => stmt.setObject(i, v)
+    }
+  }
+
+  def alignPartitions(dataFrame: DataFrame, minPartitions: Option[Int], maxPartitions: Option[Int]): DataFrame = {
```
Comment (RonBarabash) [inline src/main/scala/com/yotpo/metorikku/output/writers/jdbc/JDBCUpsertWriter.scala]: also from jdbc query

File: src/main/scala/com/yotpo/metorikku/configuration/metric/Output.scala [hunk 3]
```diff
@@ -23,10 +23,14 @@ object OutputType extends Enumeration {
   Instrumentation,
   JDBC,
   JDBCQuery,
+  JDBCUpsert,
   Elasticsearch,
   File,
   Kafka,
   Hudi = Value
```
Comment (RonBarabash) [inline src/main/scala/com/yotpo/metorikku/configuration/metric/Output.scala]: remove blank lines

**Comment by wangqinghuan** [comment]
The idea of extending JdbcQueryWriter can be summarized:
1. If  only specify query parameter, it will executes query on each dataFrame row.
2. If specify query and update paramter, it will execute query to check dataFrame row exists in database. If row exists in database, generate updateStatement.
3. If specify query and insert paramter, it will execute query to check dataFrame row exists in database. If row does not exist in database, generate insertStatement.
3. If specify query, update and insert paramter, it will execute query to check dataFrame row exists in database. If row exists in database, generate updateStatement, and if row does not exist in database, generate insertStatement.
Do you think it would be ok? @RonBarabash

**Comment by lyogev** [comment]
@wangqinghuan Are you still on this?

**Comment by lyogev** [comment]
Closing due to inactivity

**Pull request closed without merging**